### ISSUE TITLE
Improve _.isEmpty documentation relating primitive types

### ISSUE THREAD

**Issue by mcurrao** [jashkenas/underscore#2804]
I know this has already been reported way too many times. And I know it because I've faced the issue myself and found the breadcrumbs, which have lead me here. 

`_.isEmpty` behaviour by design only checks enumerable objects. Any primitive value is therefore, considered empty. 
While changing this behaviour has already been discussed (and rejected) more than once, developers have been hitting against `_.isEmpty` for years now (from at least 2012 #432 to 2018 #2788). And well... 2019 if I count myself. 

I also know this our fault, after quickly scanning through the docs we assume it will behave nicely for every variable we throw at it. 
Still, I'm sure adding an explicit example of numbers (please do NOT use 0 as an example, as it would imply it actually checks for number emptyness), and perhaps a note in bold, would help a lot of people in the future. 

Something like this would probably do:

isEmpty `_.isEmpty(object)`
----

Returns true if an enumerable object contains no values (no enumerable own-properties). For strings and array-like objects _.isEmpty checks if the length property is 0.

> _.isEmpty([1, 2, 3]);
> => false
> _.isEmpty({});
> => true

**Note that primitive types are always considered empty:**
> _.isEmpty(42);
> => true
> _.isEmpty(true);
> => true

**Comment by jgonggrijp** [comment]
@mcurrao Could you explain what is so surprising to you about the behaviour? Personally, I'd be very surprised if `_.isEmpty(42)` returned `false`.

**Comment by mcurrao** [comment]
@jgonggrijp When using variables, the concept of emptiness on a primitive-containing variable would seem to indicate the variable content not being set/filled.

Therefore:

```
let var1, var2 = 42;
_.isEmpty(var1);  // sensibly true
_.isEmpty(var2);  // expectancy unsure
```

For arrays and objects, emptyness considering content makes sense. I understand that the concept of emptyness does not apply to numbers, for example. That is why my request was not to modify its behaviour including primitives, but to modify the documentation. 
In Javascript we are used to treating any variable as a boolean variable (`if (!!var1)`). An `isEmpty` function has a tempting name for a one-fits-all solution for checking value existance, as `!!obj` would return true for an empty object, or array.
This becomes especially useful when working with unknown content format. 
```
const formValues= [ { foo: 'lala' }, { }. [ ], 42, null, null, false, [ 8 ] ];
const cleanup = (contentArr) => contentArr.filter( (elem) => _.isEmpty(elem) );
const formData = cleanup(formValues);
```
If I were to say what `cleanup` does without reading the documentation, I would say that it removes empty or null elements. I wouldn't probably expect it to remove numeric values, for example. The problem is that when debugging this example code, the documentation explains not this situation, thus making debugging the developer error (missing values) a little bit more time consuming.

**Comment by mcurrao** [comment]
Moreover, strings _are_ properly considered. I'm aware that this is due strings containing a `length` property. But this seems to contribute to the confusion, as `'42'` is not considered empty, but `42` is, both being primitives, not objects nor arrays.
As form values are serialized as either number or strings depending the used framework and the input type, `_.isEmpty` loses reliance as a form field checker. `_.isEmpty(formObj[ 'formField' ])` may return `true` or `false` depending on the type that the number was returned as.
Again, I'm not suggesting the behaviour to be updated. I'm just pointing that as strings work, and arrays, work, and objects work, there is no visible reason for which I may suspect that `_.isEmpty(formObj[ 'formField' ])` will return `false` if input value is `42`.
A simple note in the documentation stating **"Note that primitive types are always considered empty"** would really improve debugging the error, perhaps left by a previous dev.

**Comment by jgonggrijp** [comment]
So if I may rephrase this, `_.isEmpty` is meant to check whether the number of elements of a collection is nonzero (where a string is considered a collection of characters), but the name tempts people into using it on values produced from some kind of external source (such as a form field) to validate whether that source has anything nontrivial to offer.

Looking again at the documentation, I still think it is quite clear about what it will do, but I agree it is less explicit about what it will not do. I'll prepare a little pull request that adds a warning.

**Comment by jgonggrijp** [comment]
Was closed in #2840.
